### ISSUE TITLE
Missing typescript extension templates

### ISSUE THREAD

**Issue by Jonesus** [directus-community/extension-toolkit#2]
Currently we're still missing the following templates in Typescript:

- [ ] `display`
- [ ] `interface`
- [ ] `layout`
- [ ] `module`

I could use an example or a PR here as I've no experience in Vue, let alone Vue with Typescript :smile_cat:

**Comment by benhaynes** [comment]
Do we have these, just not in TS? Or not at all (as of now)? @rijkvanzanten — any thoughts on this?

**Comment by Jonesus** [comment]
I've recycled the JS examples from docs (https://docs.directus.io/guides/layouts/, https://docs.directus.io/guides/displays/, https://docs.directus.io/guides/interfaces/, https://docs.directus.io/guides/modules/) so they do work with the `-j` flag right now! But unfortunately not with Typescript...

I'm gonna learn myself some Vue eventually (and I really do wanna use Typescript with it also :stuck_out_tongue: ), but that's gonna be some time away

**Comment by rijkvanzanten** [comment]
@Jonesus This is the template you're working with now right?: https://github.com/directus-community/extension-toolkit/blob/main/src/templates/javascript/interface/src/interface.vue

A typescript alternative would be (using Vue Composition API):

```vue
<template>
	<input :value="value" @input="handleChange($event.target.value)" />
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api';

export default defineComponent({
	props: {
		value: String,
	},
	setup(props, { emit }) {
		return { handleChange };

		function handleChange(value) {
			emit('input', value);
		}
	}
});
</script>
```

@Illubyte Could you share your findings on shared packages so we don't duplicate `@vue/composition-api` all over the place?

**Comment by rijkvanzanten** [comment]
(cc @nickrum we'll have to keep an eye on Vue 2 vs Vue 3 here)

**Comment by Jonesus** [comment]
> This is the template you're working with now right?: https://github.com/directus-community/extension-toolkit/blob/main/src/templates/javascript/interface/src/interface.vue
Yeah that one, and its "parallels" in `src/templates/javascript/layout/src/layout.vue` etc. substituting `layout` twice for `module` or `display`

Besides that I'd also be interested in what would be a good `tsconfig.json` and/or `rollup.config.js`, do we use which one or both? Or would the base JS one
(https://github.com/directus-community/extension-toolkit/blob/main/src/templates/javascript/common.vue/rollup.config.js) be extensible to TS somehow, all the general project configs that TS Vue would require 😅 

Thanks for that example!

**Comment by rijkvanzanten** [comment]
@WoLfulus recently updated the tsconfig for the sub packages to be super strict, but therefore also very safe/useful. we could use that as a starting point?

```json
{
	"compilerOptions": {
		"target": "ES2015",
		"lib": ["ES2015", "DOM"],
		"module": "CommonJS",
		"moduleResolution": "node",
		"declaration": true,
		"declarationMap": true,
		"strict": true,
		"noFallthroughCasesInSwitch": true,
		"esModuleInterop": true,
		"noImplicitAny": true,
		"noImplicitThis": true,
		"noImplicitReturns": true,
		"noUnusedLocals": true,
		"noUncheckedIndexedAccess": true,
		"noUnusedParameters": true,
		"alwaysStrict": true,
		"strictNullChecks": true,
		"strictFunctionTypes": true,
		"strictBindCallApply": true,
		"strictPropertyInitialization": true,
		"resolveJsonModule": false,
		"skipLibCheck": true,
		"forceConsistentCasingInFileNames": true,
		"allowSyntheticDefaultImports": true,
		"isolatedModules": true,
		"rootDir": "./src",
		"outDir": "./dist"
	},
	"include": ["./src/**/*.ts"]
}
```

**Comment by Illubyte** [comment]
@rijkvanzanten For sure, basically in order to share packages we need to save references to imported packages to a variable on the window object. i.e.:
```javascript
// Somewhere in app/src/ in testing I added this in the plugins.ts file because it seemed logical.
// However it should be able to go anywhere.

import Vue from 'vue';
import * as CompositionApiMethods from '@vue/composition-api';

window.directus = {
  'vue': Vue,
  '@vue/composition-api': CompositionApiMethods,
};
```
The naming here isn't particularly important, however, for the sake of consistency, it makes sense to list the packages under a `directus` global variable, with property names that are a 1-to-1 match for the packages they'd be importing. 

From here, the rest of the work is on the Rollup side. 

When developing extensions, they can be developed normally, importing locally installed NPM packages as you always would. 

In the Rollup config, we need to add a plugin called `rollup-plugin-external-globals`. Rollup has an `output.globals` option, however that only compiles when the output format is `iife`, and we need to support `es`.

In the Rollup config we'd add the following:
```javascript
import externalGlobals from 'rollup-plugin-external-globals';
...
export default {
  // Other options
  plugins: [
    externalGlobals({
      'vue': 'window.directus["vue"]',
      '@vue/composition-api': 'window.directus["@vue/composition-api"]',
    })
    // Other plugins
  ]
}
```

With this setup, anywhere in our Vue project where we import `'vue'` or `'@vue/composition-api'`, upon compilation it'll replace them with `window.directus["vue"]` and `window.directus["@vue/composition-api"]`, respectively.

**Comment by Illubyte** [comment]
As a side note, I've found that this is actually necessary when using Vue 2 with the composition API plugin. I could have sworn it worked before, but for some reason the last few times I've tried compiling the `@vue/composition-api` package directly into my module, it refused to define anything reactive. I don't know what changed, or when, or if it was just a fluke with my setup, so take this with a grain of salt.

**Comment by rijkvanzanten** [comment]
@Illubyte Thanks for the heads up! Could be an initialization order problem in the app 🤔  (another thing to double check in the vue 3 migration @nickrum )
